internal/driver/glfw/loop.go

```diff
@@ -112,74 +114,81 @@ func (d *gLDriver) runGL() {
 	if d.trayStart != nil {
 		d.trayStart()
 	}
+
 	fyne.CurrentApp().Lifecycle().(*app.Lifecycle).TriggerStarted()
-	for {
-		select {
-		case <-d.done:
-			eventTick.Stop()
-			d.drawDone <- nil // wait for draw thread to stop
+
+	// Post empty event to not get stuck in the loop on first iteration.
+	PostEmptyEvent()
+
+	for range eventTick.C {
+		d.waitForEvents()
+
+		// Check if we are done with execution.
+		if atomic.LoadUint32(&d.mainDone) == 1 {
+			d.drawDone <- nil // Wait for draw thread to stop.
 			d.Terminate()
 			fyne.CurrentApp().Lifecycle().(*app.Lifecycle).TriggerStopped()
 			return
+		}
+
+		select {
 		case f := <-funcQueue:
 			f.f()
-			if f.done != nil {
-				f.done <- struct{}{}
+			f.done <- struct{}{}
+		default: // Don't get stuck waiting for event.
+		}
+
+		newWindows := []fyne.Window{}
+		reassign := false
+		for _, win := range d.windowList() {
+			w := win.(*window)
+			if w.viewport == nil {
+				continue
 			}
-		case <-eventTick.C:
-			d.tryPollEvents()
-			newWindows := []fyne.Window{}
-			reassign := false
-			for _, win := range d.windowList() {
-				w := win.(*window)
-				if w.viewport == nil {
-					continue
-				}
 
-				if w.viewport.ShouldClose() {
-					reassign = true
-					w.viewLock.Lock()
-					w.visible = false
-					v := w.viewport
-					w.viewLock.Unlock()
-
-					// remove window from window list
-					v.Destroy()
-					w.destroy(d)
-					continue
-				}
+			if w.viewport.ShouldClose() {
+				reassign = true
+				w.viewLock.Lock()
+				w.visible = false
+				v := w.viewport
+				w.viewLock.Unlock()
+
+				// remove window from window list
+				v.Destroy()
+				w.destroy(d)
+				continue
+			}
 
-				w.viewLock.RLock()
-				expand := w.shouldExpand
-				fullScreen := w.fullScreen
-				w.viewLock.RUnlock()
-
-				if expand && !fullScreen {
-					w.fitContent()
-					w.viewLock.Lock()
-					shouldExpand := w.shouldExpand
-					w.shouldExpand = false
-					view := w.viewport
-					w.viewLock.Unlock()
-					if shouldExpand {
-						view.SetSize(w.shouldWidth, w.shouldHeight)
-					}
+			w.viewLock.RLock()
+			expand := w.shouldExpand
+			fullScreen := w.fullScreen
+			w.viewLock.RUnlock()
+
+			if expand && !fullScreen {
+				w.fitContent()
+				w.viewLock.Lock()
+				shouldExpand := w.shouldExpand
+				w.shouldExpand = false
+				view := w.viewport
+				w.viewLock.Unlock()
+				if shouldExpand {
+					view.SetSize(w.shouldWidth, w.shouldHeight)
 				}
+			}
 
-				newWindows = append(newWindows, win)
+			newWindows = append(newWindows, win)
 
-				if drawOnMainThread {
-					d.drawSingleFrame()
-				}
+			if drawOnMainThread {
```

Comment:
Now that we are waiting for GLFW events this won't run 60 times a second anymore so I think we need a separate way to handle drawing on the main thread.

I'm kind of thinking out loud here but maybe the whole main loop should be a big select statement from multiple channels? 1) available GLFW events. 2) a channel signalling when a refresh is needed (because user code called .Refresh(), 3) a 60 fps animation ticker that only runs if a) we are drawing on the main thread, and b) an animation is currently in progress. So if none of these channels have available input the whole main loop can be entirely asleep in the select until something needs to be done

Reply:
I don't think a big select statement would work? I have no idea how to get the blocking `d.waitForEvents()` to send on a channel instead? We'd have to be blocked on that function (which has to run on the main thread) which makes it so that we can't get other items that are sent on the channels. That it problematic.